File: src/analysis/RDFatomistic.py
#  Copyright (C) 2012,2013,2014,2015,2016,2017,2018
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.


r"""
********************************
espressopp.analysis.RDFatomistic
********************************

Class to compute radial distribution functions in adaptive resolution simulations in subregions of the box. Can be used for regular atomistic/coarse-grained (AT/CG) adaptive resolution simulations as well as path integral-based adaptive resolution simulations. The two functions (compute, computePathIntegral) exhibit different behavior.

The regular compute function is used for regular AT/CG simulations and there are two options:

Option 1 (spanbased = True): the RDF can be calculated in a cuboid region in the center of the box (periodic in y,z, limited in x). In this case, particle pairs are considered for which at least one of them is in the defined cuboid region. This can be useful when the high resolution region has a slab geometry. No further normalization should be required.

Option 2 (spanbased = False): the routine can also calculate unnormalized RDFs using particle pairs with both particles being in the high resolution region (based on the resolution value lambda, the span parameter is not used then). This can be useful when atomistic region has complicated or spherical geometries.

In any case, only pairs of atomistic particles belonging to two different coarse-grained particles are considered. Furthermore, note that the routine uses L_y / half (L_y is the box length in y-direction) as the maximum distance for the RDF calculation, which is then binned according to rdfN during the computation. Hence, L_y should be the shortest box side (or, equally short as L_x and/or L_z).


The computePathIntegral function is used for path integral-based adaptive resolution functions. It calculates the radial distribution functions over pairs of particles between different atoms or coarse-grained beads. Note, however, that in these types of quantum/classical adaptive resolution simulations, regular coarse-grained espressopp particles are associated with each atom and the additional "AdResS" atomistic particles correspond to the different Trotter beads. This means that the routine will, for molecules consisting of multiple atoms, calculate intramolecular rdfs, averaging over the Trotter bead pairs of the ring polymers, which represent the atoms. In doing so, it considers only particles pair with matching Trotter number and with the correct atomistic types. The results are averaged over all Trotter beads. Also in this case L_y / half (L_y is the box length in y-direction) is used as the maximum distance for the RDF calculation, which is then binned according to rdfN during the computation. Furthermore, the calculation is always "spanbased" in x direction (the function ignores the spanbased flag), but in such a fashion that BOTH particles need to be in the defined cuboid region. Normalization is performed as derived in R. Potestio et al., Phys. Rev. Lett. 111, 060601 (2013), Supp. Info. This means that, considering only particles with matching Trotter numbers, the computePathIntegral function calculates the RDF between particles of type A and B within a region bounded in x-direction by :math:`X_{min}` and :math:`X_{max}` as

.. math::

    g_{slab}^{ab}(r^{AB}) = \sum_{a \in N^A} \sum_{b \in N^B} \frac{1}{ N^A N^B} \frac{\delta_\Delta(|\mathbf{r}_a - \mathbf{r}_b| - r)}{v(\mathbf{r}_a)/V_{slab}}

    \delta_\Delta(r) = \begin{cases} 1 \quad \textrm{for} \quad r<\Delta \\ 0 \quad \textrm{otherwise} \end{cases}

    v(\mathbf{r}_a)=2\pi\Delta \; r_a(2r_a-h(\mathbf{r}_a))

    h(\mathbf{r}_a) = (r_a - X^+)\theta(r_a - X^+) - (r_a - X^-)\theta(r_a - X^-)

    X^+ = X_{max} -x_a

    X^- = x_a - X_{min}

    \theta(r) = \begin{cases} 1 \quad \textrm{for} \quad r>0 \\ 0 \quad \textrm{otherwise} \end{cases}

where :math:`N^A` and :math:`N^B` are the number of particles of type A and B in the relevant subregion for the RDF calculation and :math:`V_{slab}` is the total volume of this subregion. Furthermore, :math:`r_a` denotes the radius of the spherical shell for the RDF calculation around particle :math:`a`, :math:`\Delta` is the thickness of the shell, and :math:`x_a` is the :math:`x` coordinate of particle :math:`a`. The final result is an average over all imaginary time slices (Trotter numbers).

Examples:

>>> rdf_0_1 = espressopp.analysis.RDFatomistic(system = system, type1 = 0, type2 = 1, spanbased = True, span = 1.5)
>>> # creates the class for calculating the RDF between atomistic particles of type 1 and 0 between different molecules,
>>> # At least one of these particles has to be within plus/minus 1.5 from the center of the box in x-direction

>>> rdf_0_1.compute(100)
>>> # computes the rdf using 100 bins over a distance corresponding to L_y / 2.0

.. function:: espressopp.analysis.RDFatomistic(system, type1, type2, spanbased, span)

                Constructs the RDFatomistic object.

                :param system: system object
                :param type1: type of atom 1
                :param type2: type of atom 2
                :param spanbased: (default: True) If True, calculates RDFs in a cuboid region of radius span from the center (limited in x, periodic in y,z). If False, calculates RDFs with both particles being in the high resolution region (using lambda resolution values and ignoring span parameter).
                :param span: (default: 1.0) +/- distance from centre of box in x-direction of the cuboid region used for RDF calculation if spanbased == True. If spanbased == False, this parameter is not used.
                :type system: shared_ptr<System>
                :type type1: int
                :type type2: int
                :type spanbased: bool
                :type span: real

.. function:: espressopp.analysis.RDFatomistic.compute(rdfN)

                Calculates the atomistic RDF assuming a regular atomistic/coarse-grained adaptive resolution setup.

                :param rdfN: number of bins
                :type rdfN: int
                :rtype: list of reals

.. function:: espressopp.analysis.RDFatomistic.computePathIntegral(rdfN)

                Calculates the path integral-based RDF averaging over all Trotter bead pairs with the same Trotter bead number between different ring polymers assuming a path integral-based quantum/classical adaptive resolution setup.

                :param rdfN: number of bins
                :type rdfN: int
                :rtype: list of reals
"""
from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.analysis.Observable import *
from _espressopp import analysis_RDFatomistic

class RDFatomisticLocal(ObservableLocal, analysis_RDFatomistic):

  def __init__(self, system, type1, type2, span = 1.0, spanbased = True):
    if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
      cxxinit(self, analysis_RDFatomistic, system, type1, type2, span, spanbased)

  def compute(self, rdfN):
    return self.cxxclass.compute(self, rdfN)

  def computePathIntegral(self, rdfN):
    return self.cxxclass.computePathIntegral(self, rdfN)

if pmi.isController :
  class RDFatomistic(Observable):
    __metaclass__ = pmi.Proxy
    pmiproxydefs = dict(
      pmicall = [ "compute", "computePathIntegral" ],
      cls = 'espressopp.analysis.RDFatomisticLocal'
    )

File: src/analysis/RDFatomistic.cpp
/*
  Copyright (C) 2012,2013,2014,2015,2016,2017,2018
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "python.hpp"
#include "storage/DomainDecomposition.hpp"
#include "iterator/CellListIterator.hpp"
//#include "Configuration.hpp"
#include "RDFatomistic.hpp"
#include "esutil/Error.hpp"
#include "bc/BC.hpp"

#include <boost/serialization/map.hpp>

#ifndef M_PIl
#define M_PIl 3.1415926535897932384626433832795029L
#endif

using namespace espressopp;
using namespace iterator;
using namespace std;

namespace espressopp {
  namespace analysis {

    // TODO currently works correctly if L_y is shortest side length (it's fine if L_x and/or L_z are equal to L_y)
    // rdfN is a level of discretisation of rdf (how many elements it contains)
    python::list RDFatomistic::computeArray(int rdfN) const {

      System& system = getSystemRef();
      esutil::Error err(system.comm);
      Real3D Li = system.bc->getBoxL();
      Real3D Li_half = Li / 2.;

      int nprocs = system.comm->size();
      int myrank = system.comm->rank();

      real histogram [rdfN];
      for(int i=0; i<rdfN; i++) histogram[i]=0;

      real dr = Li_half[1] / (real)rdfN; // If you work with nonuniform Lx, Ly, Lz, you
      // should use for Li_half[XXX] the shortest side length. Currently using L_y

      int num_part = 0;
      map< size_t, data > config;
      for (int rank_i=0; rank_i<nprocs; rank_i++) {
        map< size_t, data > conf;
        if (rank_i == myrank) {
          shared_ptr<FixedTupleListAdress> fixedtupleList = system.storage->getFixedTuples();
          CellList realCells = system.storage->getRealCells();
          for(CellListIterator cit(realCells); !cit.isDone(); ++cit) {

            Particle &vp = *cit;
            FixedTupleListAdress::iterator it2;
            it2 = fixedtupleList->find(&vp);

            if (it2 != fixedtupleList->end()) {  // Are there atomistic particles for given CG particle? If yes, use those for calculation.
              std::vector<Particle*> atList;
              atList = it2->second;
              for (std::vector<Particle*>::iterator it3 = atList.begin();
                   it3 != atList.end(); ++it3) {
                Particle &at = **it3;
                int id = at.id();
                data tmp = { at.position(), at.type(), vp.id(), vp.lambda() };
                conf[id] = tmp;
              }
            }

            else {  // If not, use CG particle itself for calculation.
              std::cout << "WARNING! In RDFatomistic, no atomistic AdResS particle found!\n";
              exit(1);
            }

          }
        }

        boost::mpi::broadcast(*system.comm, conf, rank_i);

        // for simplicity we will number the particles from 0
        for (map<size_t,data>::iterator itr=conf.begin(); itr != conf.end(); ++itr) {
          data p = itr->second;
          config[num_part] = p;
          num_part ++;
        }
      }

      int num_pairs = 0;
      // now all CPUs have all particle coords and num_part is the total number of particles

      // use all cpus
      // TODO it could be a problem if   n_nodes > num_part
      int numi = num_part / nprocs + 1;
      int mini = myrank * numi;
      int maxi = mini + numi;

      if(maxi>num_part) maxi = num_part;

      int perc=0, perc1=0;
      for(int i = mini; i<maxi; i++) {
        Real3D coordP1 = config[i].pos;
        int typeP1 = config[i].type;
        int molP1 = config[i].molecule;
        real resolutionP1 = config[i].resolution;


        if (spanbased == true) {
          if((coordP1[0] < Li_half[0]+span) && (coordP1[0] > Li_half[0]-span) ) {
            for(int j = i+1; j<num_part; j++) {
              Real3D coordP2 = config[j].pos;
              int typeP2 = config[j].type;
              int molP2 = config[j].molecule;

              if (molP1 != molP2) {

                if( ( (typeP1 == target1) && (typeP2 == target2) ) || ( (typeP1 == target2) && (typeP2 == target1) ) ) {
                  Real3D distVector = (0.0, 0.0, 0.0);
                  system.bc->getMinimumImageVector(distVector, coordP1, coordP2);

                  int bin = (int)( distVector.abs() / dr );
                  if( bin < rdfN) {
                    histogram[bin] += 1.0;
                  }
                  num_pairs += 1;

                }
              }
            }
          }
        }
        else {
          if( resolutionP1 == 1.0 ) {
            for(int j = i+1; j<num_part; j++) {
              Real3D coordP2 = config[j].pos;
              int typeP2 = config[j].type;
              int molP2 = config[j].molecule;
              real resolutionP2 = config[j].resolution;

              if (molP1 != molP2) {

                if( ( (typeP1 == target1) && (typeP2 == target2) ) || ( (typeP1 == target2) && (typeP2 == target1) ) ) {

                  num_pairs += 1;

                  if( resolutionP2 == 1.0 ) {
                    Real3D distVector = (0.0, 0.0, 0.0);
                    system.bc->getMinimumImageVector(distVector, coordP1, coordP2);

                    int bin = (int)( distVector.abs() / dr );
                    if( bin < rdfN) {
                      histogram[bin] += 1.0;
                    }
                  }

                }
              }
            }
          }
        }
      }

      real totHistogram[rdfN];
      boost::mpi::all_reduce(*system.comm, histogram, rdfN, totHistogram, plus<real>());

      int tot_num_pairs = 0;
      boost::mpi::all_reduce(*system.comm, num_pairs, tot_num_pairs, plus<int>());
      int nconfigs = 1;
      real rho = (real)1.0 / (Li[0]*Li[1]*Li[2]);
      real factor = 4.0 * M_PIl * dr * rho * (real)nconfigs * (real)tot_num_pairs;

      for(int i=0; i < rdfN; i++) {
        real radius = (i + 0.5) * dr;
        totHistogram[i] /= factor * (radius*radius + dr*dr / 12.0);
      }

      python::list pyli;
      for(int i=0; i < rdfN; i++) {
        pyli.append( totHistogram[i] );
      }
      return pyli;
    }

    python::list RDFatomistic::computeArrayPathIntegral(int rdfN) const {

      System& system = getSystemRef();
      esutil::Error err(system.comm);
      Real3D Li = system.bc->getBoxL();
      Real3D Li_half = Li / 2.;

      real Xmax = Li_half[0] + span;
      real Xmin = Li_half[0] - span;

      int nprocs = system.comm->size();
      int myrank = system.comm->rank();

      real histogram [rdfN];
      for(int i=0; i<rdfN; i++) histogram[i]=0;

      real dr = Li_half[1] / (real)rdfN; // If you work with nonuniform Lx, Ly, Lz, you
      // should use for Li_half[XXX] the shortest side length. Currently using L_y

      int num_part = 0;
      map< size_t, dataPathIntegral > config;
      for (int rank_i=0; rank_i<nprocs; rank_i++) {
        map< size_t, dataPathIntegral > conf;
        if (rank_i == myrank) {
          shared_ptr<FixedTupleListAdress> fixedtupleList = system.storage->getFixedTuples();
          CellList realCells = system.storage->getRealCells();
          for(CellListIterator cit(realCells); !cit.isDone(); ++cit) {

            Particle &vp = *cit;
            FixedTupleListAdress::iterator it2;
            it2 = fixedtupleList->find(&vp);

            if (it2 != fixedtupleList->end()) {
              std::vector<Particle*> atList;
              atList = it2->second;
              for (std::vector<Particle*>::iterator it3 = atList.begin();
                   it3 != atList.end(); ++it3) {
                Particle &at = **it3;

                if ( (at.position()[0] < Xmax) && (at.position()[0] > Xmin) ) {
                  int id = at.id();
                  dataPathIntegral tmp = { at.position(), at.type(), vp.id(), at.pib() };
                  conf[id] = tmp;
                }
              }
            }

            else {  // If not, use CG particle itself for calculation.
              std::cout << "WARNING! In RDFpathintegral, no atomistic AdResS particle found!\n";
              exit(1);
            }

          }
        }

        boost::mpi::broadcast(*system.comm, conf, rank_i);

        // for simplicity we will number the particles from 0
        for (map<size_t, dataPathIntegral>::iterator itr=conf.begin(); itr != conf.end(); ++itr) {
          dataPathIntegral p = itr->second;
          config[num_part] = p;
          num_part ++;
        }
      }

      int num_pairs = 0;

      int numi = num_part / nprocs + 1;
      int mini = myrank * numi;
      int maxi = mini + numi;

      if(maxi>num_part) maxi = num_part;

      int perc=0, perc1=0;
      for(int i = mini; i<maxi; i++) {
        Real3D coordP1 = config[i].pos;
        int typeP1 = config[i].type;
        int molP1 = config[i].molecule;
        int pibP1 = config[i].pib;

        if( (coordP1[0] < Xmax) && (coordP1[0] > Xmin) ) {
          for(int j = i+1; j<num_part; j++) {
            Real3D coordP2 = config[j].pos;
            int typeP2 = config[j].type;
            int molP2 = config[j].molecule;
            int pibP2 = config[j].pib;

            if ( (molP1 != molP2) && (pibP1 == pibP2) && (coordP2[0] < Xmax) && (coordP2[0] > Xmin) ) {

              if( ( (typeP1 == target1) && (typeP2 == target2) ) || ( (typeP1 == target2) && (typeP2 == target1) ) ) {

                Real3D distVector = coordP1 - coordP2;

                // minimize the distance in simulation box
                for(int ii=0; ii<3; ii++) {
                  if( distVector[ii] < -Li_half[ii] ) distVector[ii] += Li[ii];
                  if( distVector[ii] >  Li_half[ii] ) distVector[ii] -= Li[ii];
                }

                real absdist = distVector.abs();

                int bin = (int)( absdist / dr );
                if( bin < rdfN) {

                  // Normalization (see R. Potestio et al., Phys. Rev. Lett. 111, 060601 (2013), Supp. Info.)
                  real Xplus = Xmax - coordP1[0];
                  real Xminus = coordP1[0] - Xmin;

                  real h_norm = 0.0;

                  if (Xplus <= absdist) {
                    h_norm += absdist - Xplus;
                  }
                  if (Xminus <= absdist) {
                    h_norm += absdist - Xminus;
                  }

                  real v_norm = 1.0/(2.0*M_PIl*dr*absdist*(2.0*absdist - h_norm));

                  histogram[bin] += v_norm;
                }
                num_pairs += 1;

              }
            }
          }
        }

      }

      real totHistogram[rdfN];
      boost::mpi::all_reduce(*system.comm, histogram, rdfN, totHistogram, plus<real>());

      int tot_num_pairs = 0;
      boost::mpi::all_reduce(*system.comm, num_pairs, tot_num_pairs, plus<int>());

      // Normalization (see R. Potestio et al., Phys. Rev. Lett. 111, 060601 (2013), Supp. Info.)
      real subvol = 2.0*Li[1]*Li[2]*span;
      if(span > 0.5*Li[0]) {
        subvol = Li[1]*Li[2]*Li[0];
      }
      real rho = (real)1.0 / (Li[0]*Li[1]*Li[2]);
      for(int i=0; i < rdfN; i++) {
        totHistogram[i] /= ( (real)tot_num_pairs / subvol );
      }
      // Normalization

      python::list pyli;
      for(int i=0; i < rdfN; i++) {
        pyli.append( totHistogram[i] );
      }
      return pyli;
    }

    // TODO: this dummy routine is still needed as we have not yet ObservableVector
    real RDFatomistic::compute() const {
      return -1.0;
    }

    void RDFatomistic::registerPython() {
      using namespace espressopp::python;
      class_<RDFatomistic, bases< Observable > >
      ("analysis_RDFatomistic", init< shared_ptr< System >, int, int, real, bool >())
      .def("compute", &RDFatomistic::computeArray)
      .def("computePathIntegral", &RDFatomistic::computeArrayPathIntegral)
      ;
    }
  }
}
